### PR TITLE
[tt-crank] compile lowering for fused SDPA training

### PR DESCRIPTION
**Pull request by agobeljicTT**
Stacked on #9348. Third of the SDPA training stack (split of #9323): `torch.compile` path.

The fused overrideable forward lowers to `sdpa_fw` when its logsumexp slot is read (a backward follows) and to the prefill op otherwise; the backward lowers to `sdpa_bw`. `_needs_node` lets a lowering see its FX node for that slot check. The eager choice stub no longer forces MATH while tracing. `OPT_LEVEL 1` promotes the composites to `ttnn.sdpa_fw/bw`, `OPT_LEVEL 0` inlines the decomposition; both tested.

Tests: `test_sdpa.py` compile modes (fused, decompose, dropout, f32, TP/DP), two-program artifact check (no host round-trip), lse tile layout, `models/test_attention_training.py`. OPT 1 cases `xfail(run=False)` under ttsim (OPT 1 aborts there).

### CODE REVIEW

File: tt-crank/python/tt_crank/torch/_compile.py [hunk 1]
```diff
@@ -126,6 +126,20 @@ def decorator(fn):
...
+    fn._needs_node = True
+    return fn
+
+
+def _used_output_slots(node) -> set[int]:
+    return {
+        u.args[1]
+        for u in node.users
+        if u.op == "call_function" and u.target is operator.getitem
+    }
```
[the remark this answers is not in the document]
Comment (agobeljicTT) [reply tt-crank/python/tt_crank/torch/_compile.py:140]: dynamo unpacks a tuple slice into per-element getitem nodes with int indices before the graph reaches the backend, a slice never shows up as a getitem arg on an aten tuple output. test_sdpa_fw_logsumexp_matches_torch uses exactly [:2] and asserts ttnn.sdpa_fw is in the ir, it passes.

File: tt-crank/tests/python/test_sdpa.py [hunk 2]
```diff
@@ -384,3 +416,75 @@ def test_sdpa_eager_multi_chip_causal_backward(tt_pg, parallel: str) -> None:
...
+    q, k, v = (torch.randn(_B, _H, _S, _E, dtype=_DT) for _ in range(3))
+    logits = (q.float() @ k.float().transpose(2, 3)) * _E**-0.5
+    ref_lse = logits.masked_fill(
+        ~torch.ones(_S, _S, dtype=torch.bool).tril(), float("-inf")
+    ).logsumexp(-1)
+
+    def fw(a, b, c):
+        return torch.ops.aten._scaled_dot_product_fused_attention_overrideable(
+            a, b, c, None, 0.0, True
+        )[:2]
```
[the remark this answers is not in the document]
Comment (agobeljicTT) [reply tt-crank/tests/python/test_sdpa.py:590]: same as above, the slice is already two int getitems by the time the lowering runs and the test passes as written.

File: tt-crank/tests/python/models/test_attention_training.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        batch, seq, model_dim = x.shape
+        q, k, v = (
+            t.view(batch, seq, self.num_heads, -1).transpose(1, 2)
+            for t in self.qkv(x).chunk(3, dim=-1)
+        )
+        attn = F.scaled_dot_product_attention(q, k, v, is_causal=True)
+        return self.proj(attn.transpose(1, 2).reshape(batch, seq, model_dim))
+
+
+def _train(model: nn.Module, x: torch.Tensor, y: torch.Tensor) -> list[float]:
```
Comment (vladimirjovanovicTT) [inline tt-crank/tests/python/models/test_attention_training.py:38]: Having _train helper here (and in other tests) doesn't scale. I think we need a shared train helper.
I am not convinced that we want loops in module-level tests, I see this need in a model-level test. Essentially here we are creating a fake model-level test for a module, I think we do not want this.

**Review by vladimirjovanovicTT — CHANGES_REQUESTED** [review]
